src/alphafold_impact/pipelines/data_analysis_nih_funding_opportunities/nodes.py

```diff
@@ -0,0 +1,146 @@
+"""
+This module contains functions for processing funding opportunities and
+related MeSH data.
+"""
+
+import pandas as pd
+from typing import List
+
+
+def _find_lower_lvl_mesh_terms(
+    full_mesh_tree: pd.DataFrame, tree_numbers: list
+) -> pd.DataFrame:
+    """
+    Filter the DataFrame by rows where the 'tree_number' column starts with any of
+    the strings given in tree_numbers.
+
+    Args:
+        full_mesh_tree: A pandas DataFrame with a 'tree_number' column.
+        tree_numbers: A string to match the start of the 'tree_number' entries.
+
+    Returns:
+        pd.DataFrame: A DataFrame containing the rows where 'tree_number' starts
+            with the given string.
+    """
+    return full_mesh_tree[
+        full_mesh_tree["tree_number"].apply(
+            lambda x: any(
+                x.startswith(tree_number) for tree_number in tree_numbers if pd.notna(x)
+            )
+        )
+    ]
+
+
+def _list_mesh_duis(
+    subfield_mesh_dict: dict, full_mesh_tree: pd.DataFrame
+) -> List[str]:
+    """
+    Return list of MeSH DUIs
+
+    Args:
+        subfield_mesh_dict (dict): Dictionary with keys of MeSH names and
+            list of MeSH tree numbers for the values
+        full_mesh_tree (pd.DataFrame): Full collection of MeSH names,
+            DUIs and tree numbers
+
+    Returns:
+        list: A list of MeSH DUIs
+    """
+    return [
+        DUI
+        for tree_numbers in subfield_mesh_dict.values()
+        for DUI in _find_lower_lvl_mesh_terms(full_mesh_tree, tree_numbers).DUI.values
+    ]
+
+
+def _search_text_in_multiple_cols(
+    df: pd.DataFrame, columns: list[str], search_terms: list[str]
+) -> pd.DataFrame:
+    """
+    Search for any of several texts in any of multiple specified columns of
+    a pandas DataFrame and return rows containing those texts, ignoring case
+    differences.
+
+    Args:
+        df (pd.DataFrame): The DataFrame to search in.
+        columns (list[str]): The names of the columns to perform the search.
+        search_terms (list[str]): The terms to search for.
+
+    Returns:
+        pd.DataFrame: A DataFrame containing the rows where any text is found
+            in any column, case-insensitively.
+    """
+    matching_rows = pd.DataFrame()
+    for column in columns:
```

Comment:
I am able to match 2 opps with "crystallography" (that is, not using x-ray) which feel relevant, but your keywords don't include. Both of these are captured via the DUI, but just saying that we might be missing others.